config/src/main/java/com/typesafe/config/impl/ConfigDocumentParser.java

```diff
@@ -0,0 +1,620 @@
+
+        // In arrays and objects, comma can be omitted
+        // as long as there's at least one newline instead.
+        // this skips any newlines in front of a comma,
+        // skips the comma, and returns true if it found
+        // either a newline or a comma. The iterator
+        // is left just after the comma or the newline.
+        private boolean checkElementSeparator(Collection<AbstractConfigNode> nodes) {
+            if (flavor == ConfigSyntax.JSON) {
+                Token t = nextTokenIgnoringWhitespace(nodes);
+                if (t == Tokens.COMMA) {
+                    nodes.add(new ConfigNodeSingleToken(t));
+                    return true;
+                } else {
+                    putBack(t);
+                    return false;
+                }
+            } else {
+                boolean sawSeparatorOrNewline = false;
+                Token t = nextToken();
+                while (true) {
+                    if (Tokens.isIgnoredWhitespace(t) || isUnquotedWhitespace(t)) {
+                        //do nothing
+                    } else if (Tokens.isNewline(t)) {
+                        // newline number is the line just ended, so add one
+                        lineNumber = t.lineNumber() + 1;
+                        sawSeparatorOrNewline = true;
+
+                        // we want to continue to also eat
+                        // a comma if there is one.
+                    } else if (t == Tokens.COMMA) {
+                        nodes.add(new ConfigNodeSingleToken(t));
+                        return true;
+                    } else {
+                        // non-newline-or-comma
+                        putBack(t);
+                        return sawSeparatorOrNewline;
+                    }
+                    nodes.add(new ConfigNodeSingleToken(t));
+                    t = nextToken();
+                }
+            }
+        }
+
+        // merge a bunch of adjacent values into one
+        // value
+        private ConfigNodeComplexValue parseConcatenation(Collection<AbstractConfigNode> nodes) {
+            // this trick is not done in JSON
+            if (flavor == ConfigSyntax.JSON)
+                return null;
+
+            // create only if we have value tokens
+            ArrayList<AbstractConfigNode> values = new ArrayList<AbstractConfigNode>();
+            int valueCount = 0;
+
+            // ignore a newline up front
+            Token t = nextTokenIgnoringWhitespace(nodes);
+            while (true) {
+                AbstractConfigNodeValue v = null;
+                if (Tokens.isValue(t) || Tokens.isUnquotedText(t)
+                        || Tokens.isSubstitution(t) || t == Tokens.OPEN_CURLY
+                        || t == Tokens.OPEN_SQUARE) {
+                    // there may be newlines _within_ the objects and arrays
+                    v = parseValue(t);
+                    valueCount++;
+                } else if (Tokens.isIgnoredWhitespace(t)) {
```

Comment:
I'm planning to refactor this a bit before I work on the `ConfigDocument` stuff, since if it sees a complex value like an object or an array, if there is no concatenation, it will throw out the value it parsed, which could be a large performance hit in the cases of very large objects or arrays.